src/Xamarin.Android.Build.Tasks/Utilities/NativeAssemblyGenerator/NativeAssemblyGenerator.cs

```diff
@@ -0,0 +1,1223 @@
+			if (type == typeof(byte) || type == typeof(sbyte)) {
+				status.TempWriter.GetStringBuilder ().Clear ();
+
+				WriteData (status.TempWriter, value, hex, comment, useBlockComment);
+				AddStructureField (status, GetNativeType<T> (), status.TempWriter.ToString (), (uint)value.Length);
+				return;
+			}
+
+			bool first = true;
+			foreach (T v in value) {
+				WriteData (status, v, hex, first ? comment : null, useBlockComment);
+				if (first) {
+					first = false;
+				}
+			}
+		}
+
+		void AddStructureField (StructureWriteContext status, NativeType nativeType, string contents, uint arrayElements = 0)
+		{
+			// Structure alignment is always the same as its most strictly aligned component, as per
+			//
+			// https://github.com/ARM-software/abi-aa/blob/320a56971fdcba282b7001cf4b84abb4fd993131/aapcs64/aapcs64.rst#59composite-types
+			// https://github.com/ARM-software/abi-aa/blob/320a56971fdcba282b7001cf4b84abb4fd993131/aapcs32/aapcs32.rst#53composite-types
+			if (nativeType.Alignment > status.ByteAlignment) {
+				status.ByteAlignment = nativeType.Alignment;
+			}
+
+			if (arrayElements > 0) {
+				status.Size += nativeType.Size * arrayElements;
+			} else {
+				status.Size += nativeType.Size;
+			}
+			status.Fields.Add (new StructureField (nativeType, contents, arrayElements));
+		}
+
+		public virtual uint WriteData<T> (T value, bool hex = true, string? comment = null, bool useBlockComment = false) where T: struct
+		{
+			return WriteData<T> (Output, value, hex, comment, useBlockComment);
+		}
+
+		public void WriteDataSection (TextWriter writer, string? name = null, bool writable = true)
+		{
+			string sectionName = writable ? ".data" : ".rodata";
+			if (!String.IsNullOrEmpty (name)) {
+				sectionName = $"{sectionName}.{name}";
+			}
+
+			SectionFlags flags = SectionFlags.Allocatable;
+			if (writable) {
+				flags |= SectionFlags.Writable;
+			}
+
+			WriteSection (writer, sectionName, flags, SectionType.Data);
+		}
+
+		public void WriteDataSection (string? name = null, bool writable = true)
+		{
+			WriteDataSection (Output, name, writable);
+		}
+
+		public void WriteStringSection (TextWriter writer, string name)
+		{
+			WriteSection (
+				writer,
+				$".rodata.{name}",
+				SectionFlags.Allocatable | SectionFlags.HasCStrings | SectionFlags.Mergeable,
+				SectionType.Data,
+				flagSpecificArgs: "1"
+			);
+		}
+
+		public void WriteStringSection (string name)
+		{
+			WriteStringSection (Output, name);
+		}
+
+		public void WriteSection (TextWriter writer, string name, SectionFlags flags = SectionFlags.None, SectionType type = SectionType.None, string? flagSpecificArgs = null, string? numberOrCustomFlag = null, string? numberOrCustomType = null)
+		{
+			writer.WriteLine ();
+			writer.Write ($"{Indent}.section{Indent}{name}");
+			if (flags != SectionFlags.None) {
+				writer.Write (", \"");
+
+				var list = new StringBuilder ();
+				if ((flags & SectionFlags.Allocatable) == SectionFlags.Allocatable) {
```

Comment:
Have you considered instead using `if (flags.HasFlag(SectionFlags.Allocatable)) {…}`?

Reply:
Never liked it... an unnecessary function call. Will change.